experimental/offline_attestation/server/src/main.rs

```diff
@@ -0,0 +1,143 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#![feature(try_blocks)]
+
+use anyhow::anyhow;
+use clap::Parser;
+use log::{debug, info};
+use offline_attestation_shared::{
+    decrypt, encrypt, generate_private_key, AttestationReport, EncryptedRequest, EncryptedResponse,
+    Handle, PublicKeyInfo,
+};
+use std::{
+    net::{Ipv6Addr, SocketAddr},
+    sync::Arc,
+};
+use warp::{hyper::StatusCode, reply::Response, Filter, Reply};
+
+const ECHO_PATH: &str = "encrypted_echo";
+
+const PUBLIC_KEY_INFO_PATH: &str = "public_key_info";
+
+#[derive(Parser, Clone, Debug)]
+#[clap(about = "Offline Attestation Server")]
+pub struct Opt {
+    #[clap(long, default_value = "8080", help = "Port number on which to listen.")]
+    port: u16,
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    env_logger::init();
+    offline_attestation_shared::init();
+    let opt = Opt::parse();
+    let address = SocketAddr::from((Ipv6Addr::UNSPECIFIED, opt.port));
+
+    // Generate a private key for the lifetime of the server at startup.
+    // In a production implementation there would typically be some form of key provisioning
+    // mechanism that will ensure all servers that run the right code in an appropriate enclave
+    // receive the same key.
+    debug!("Generating private key");
+    let private_key_handle = Arc::new(generate_private_key()?);
```

Comment:
Just to confirm, is the same private key shared between multiple backend servers?

Reply:
Yes, it will be in the case where the runtime supports multiple servers sharing keys.